cms/admin/pageadmin.py

```diff
@@ -841,28 +897,69 @@ def copy_page(self, request, page_id, extra_context=None):
     @require_POST
     @transaction.atomic
     def publish_page(self, request, page_id, language):
+        all_published = True
+
         try:
-            page = Page.objects.get(id=page_id, publisher_is_draft=True)
+            page = Page.objects.get(
+                id=page_id,
+                title_set__language=language,
+                publisher_is_draft=True,
+            )
         except Page.DoesNotExist:
             page = None
-        # ensure user has permissions to publish this page
-        all_published = True
-        if page:
-            if not self.has_publish_permission(request, obj=page):
-                return HttpResponseForbidden(force_text(_("You do not have permission to publish this page")))
-            published = page.publish(language)
-            if not published:
-                all_published = False
+
         statics = request.GET.get('statics', '')
+
         if not statics and not page:
-            raise Http404("No page or stack found for publishing.")
+            raise Http404("No page or static placeholder found for publishing.")
+
+        # ensure user has permissions to publish this page
+        if page and not self.has_publish_permission(request, obj=page):
+            return HttpResponseForbidden(force_text(_("You do not have permission to publish this page")))
+
+        if page:
+            operation_token = self._send_pre_page_operation(
+                request,
+                operation=operations.PUBLISH_PAGE_TRANSLATION,
+                obj=page,
+                translation=page.get_title_obj(language=language),
+            )
+            all_published = page.publish(language)
+            page = page.reload()
+            self._send_post_page_operation(
+                request,
+                operation=operations.PUBLISH_PAGE_TRANSLATION,
+                token=operation_token,
+                obj=page,
+                translation=page.get_title_obj(language=language),
+            )
+
         if statics:
-            static_ids = statics .split(',')
-            for pk in static_ids:
-                static_placeholder = StaticPlaceholder.objects.get(pk=pk)
+            static_ids = statics.split(',')
+            static_placeholders = StaticPlaceholder.objects.filter(pk__in=static_ids)
+
+            for static_placeholder in static_placeholders.iterator():
+                # TODO: Maybe only send one signal...
+                # this would break the obj signal format though
```

Comment:
It looks ugly, but emitting one signal per static placeholder it's the only sane option